Refactoring of `DateTime` related functions

larshinueber opened the pull request:
This PR makes the tudat-sided changes to facilitate https://github.com/tudat-team/tudatpy/pull/304.

This includes:
- Move `convertYearAndDaysInYearToTudatDate`, `timePointToDateTime`, `dateTimeToTimePoint` function definitions from tudatpy to tudat
- Move all `DateTime` related functions inside `DateTime` class
  - All functions that create a `DateTime` object are defined as static functions named `from*` (`DateTime::fromTimePoint`, `fromYearAndDaysInYear` etc.)
  - Functions that create a new `DateTime` object based on the current instance are added as instance methods (`addSecondsToDateTime` and `addDaysToDateTime`)
- Deprecation warning is added to functions that are still needed for backwards compatibility

Remaining TODOs:
- [ ] Add tests for functions that were ported over from tudatpy

larshinueber committed:
refactor tudat<->chrono functions

larshinueber committed:
add remaining DateTime conversion methods

larshinueber committed:
Move remaining functions to DateTime class

DominicDirkx commented:
Very nice! Once the small extra unit tests are this, we can merge this!

larshinueber committed:
add DateTime tests

larshinueber commented:
I've added tests for the new functionality but run into a few issues:
- Originally I was thinking to compare the resulting Tudat DateTime object after converting `DateTime->std::chrono->DateTime`. This doesn't work with the test data defined in `testDateTimeConversions` for cases where the seconds component is close to 60 seconds, as they are then rounded up in the `DateTime::timePoint` method (see [dateTime.h#L235](https://github.com/larshinueber/tudat/blob/48d7fb3394ce93283de13d821889e804cb65ebb5/include/tudat/astro/basic_astro/dateTime.h#L235)), which overflows to minute/hour components.
- When comparing the epochs of the DateTime objects after converting `DateTime->std::chrono->DateTime`, the test fails for dates outside the years 1678-2262 ([unitTestDateTime.cpp#L144](https://github.com/larshinueber/tudat/blob/ee026075daf694e879fc5ff05c45c8938389ef63/tests/src/astro/basic_astro/unitTestDateTime.cpp#L144)). This seems to be due to `std::chrono::duration` being represented as nanoseconds using a signed 64 bit integer. For now I've marked the tests as expected failing and I would propose to open a separate issue for this.

@DominicDirkx let me know if this is sufficient for now!

DominicDirkx commented:
Hi Lars, again thanks for putting in all this work to make this code so much better and more consistent. I'm pretty sure any conversion we do through `std::chrono` will cause us to lose some numerical resolution. If I understand correctly, the issues you raise are only associated with conversions to/from std::chrono timePoint objects right? Correct me if I'm wrong, but these are not used as 'intermediate' functions anywhere right? But, since pybind maps python.datetime to std::chrono timepoints, these conversions do matter for our interfaces (?)

If all my assumptions here are right, I think your second point is quite easily addressed: if the conversion is queried outside the range 1678-2262 , an exception is thrown.

For the first point (issue due to overflow if seconds is close to 60), what happens to teh resulting conversion, is the error still relatively small, or does the overflow cause the error to blow up beyond reasonable bounds?

DominicDirkx commented:
I played around a little bit, and it seems that the total error in terms of epoch remains very small, but that the numerical representation in separate numbers for hours, minutes, seconds becomes big (?)

I think the best way to approach this is:
* Modify the unit tests to only perform the tests that are expected to pass (e.g. with an if statement on the test index) and, where possible, check whether the other cases give the 'expected' abnormal behaviour (e.g. 11:59:59.99999999999==12:00:00 or something similar)

While playing, I also found a bug in the DateTime::isoString function. The line:

std::string fractionalSecondString =
                utilities::to_string_with_precision< long double >( fractionalSeconds, numberOfFractionalSecondDigits );

incorrectly makes 0.99999999999999999 into 1.0, which turns 59.99999999999999999 into 59.0 when printing as an iso string. This is not necesarilly something that has to be tackled now, but if we don't we should open a separate issue for it.

DominicDirkx commented:
Finally, it seems like the test is failing on windows, crashing as soon as it hits the test for 1910. Perhaps the Windows time span is different than Mac or linux?

larshinueber commented:
Thanks @DominicDirkx for the detailed reply!

> If I understand correctly, the issues you raise are only associated with conversions to/from std::chrono timePoint objects right? Correct me if I'm wrong, but these are not used as 'intermediate' functions anywhere right? But, since pybind maps python.datetime to std::chrono timepoints, these conversions do matter for our interfaces (?)

This is correct, the functions from and to timePoint objects are only used for exposures to the user, not inside tudat. Having this is useful e.g. if you would like to plot a timeseries, as you can pass simply pass a list of Python `datetime` objects to matplotlib, which takes care of formatting the epoch nicely depending on the timescales in the plot:

<details><summary>Plotting example</summary>
<p>

```python
import matplotlib.pyplot as plt
from tudatpy.astro.time_conversion import DateTime

fig, ax = plt.subplots()

# dts will be automatically formatted as years by matplotlib
dts = [DateTime.from_year_and_day_of_year(2000+i, 1).to_python_datetime() for i in range(10)]
# dummy data
ys = [i for i in range(10)]

ax.plot(dts, ys)
ax.set_xlabel("Date")
ax.set_ylabel("Dummy data")
plt.show()
```

![image](https://github.com/user-attachments/assets/47fed726-6a8e-4d68-8454-45d04e76bcc6)

</p>
</details> 

> If all my assumptions here are right, I think your second point is quite easily addressed: if the conversion is queried outside the range 1678-2262 , an exception is thrown.

I think this would work in the direction Tudat DateTime -> std::chrono but not the other way around (please correct me if I'm wrong!). I might be missing something in the C++ docs but from my understanding the `std::chrono::system_clock::time_point` instance represents a time duration relative to some epoch (1900-01-01 in this case) so it's not possible to distinguish a duration that is overflown from a "correct" duration inside the range that can be represented?
I dont know if there is any way around this (can the time representation of the input `std::chrono::system_clock::time_point datetime` be "forced" to only microsecond resolution?) or if the user should just be explicitly warned about the limitation.

> I played around a little bit, and it seems that the total error in terms of epoch remains very small, but that the numerical representation in separate numbers for hours, minutes, seconds becomes big (?)

Exactly 👍 I'll implement the test cases depending on the index.

> While playing, I also found a bug in the DateTime::isoString function

Interesting, I hadn't noticed this. When called from tudatpy, 59.99999999999999999 is turned into 60.0 in the iso string. I would open a separate issue for this 👍 

> Finally, it seems like the test is failing on windows, crashing as soon as it hits the test for 1910. Perhaps the Windows time span is different than Mac or linux?

Yeah I found this one odd. The test checks if an error is thrown for invalid dates (which it is incorrectly not on Windows) but the PR doesn't change any of the DateTime constructor or validation code.. Will look further into this

larshinueber committed:
add expected offsets to date/time components

DominicDirkx committed:
Adding debugging output to test Windows test failure

DominicDirkx committed:
Addind some more test output

DominicDirkx committed:
Added test output to datetime

DominicDirkx commented:
I've been adding some error output to the DateTime code, to nail down where the Windows and Mac failures come from. My guess it's in the    ` static DateTime fromTime( const TimeType &timeInput )` when calling `DateTime reconstructedDateTime = DateTime::fromTime( currentTime );` in the test.

larshinueber committed:
split datetime test cases

larshinueber committed:
check if DateTime epoch is out of range for conversion to timepoint

larshinueber committed:
test for exception when constructing timepoint

DominicDirkx committed:
Turning off build of most tests; adding test output

DominicDirkx committed:
Adding more error output

DominicDirkx committed:
Corrected error (?)

larshinueber committed:
make timepoint tests dependent on min/max chrono representable epoch

larshinueber committed:
reenable tests

DominicDirkx commented:
Tests passing! I'll have a look in the coming days, but I think we can merge this now

larshinueber committed:
fix lower representation count

larshinueber committed:
add debugging outputs, disable unrelated tests

larshinueber committed:
add more test debugging statements

larshinueber committed:
add local tm output before datetime construction

larshinueber committed:
microsecond count output

larshinueber committed:
try to make chrono limits dependent on duration representation

larshinueber commented:
I'm afraid the passing tests were an implementation error on my side 😅
I tried another approach in finding the limits representation, but also this one doesn't seem work. I suspect that is because `std::time_t` and `std::tm` are stored in different representations than `std::chrono::system_clock::time_point`. They will fail in the construction of a DateTime object from a timePoint, while the other way around still works perfectly fine.

DominicDirkx commented:
Thanks for continuing to dive into this! So, the issue is a (system-dependent) range of validity of the time periods the types can cover right? So, the creation of the DateTime from a time_point will fail on certain epochs. Is the remaining problem to have the unit tests capture this correctly, or is there something (as far as you know) in the implementation that is missing?

larshinueber commented:
> So, the issue is a (system-dependent) range of validity of the time periods the types can cover right? So, the creation of the DateTime from a time_point will fail on certain epochs.

Yes, this is correct 👍

> Is the remaining problem to have the unit tests capture this correctly, or is there something (as far as you know) in the implementation that is missing?

I suspect this is not only an issue in the unit tests but also when creating a time point object, since they use the same function in determining the valid range for conversion. The Linux approach in finding this range doesn't seem to work on MacOS and Windows. I might be missing something here?

To be quite honest I am running a bit out of ideas how this can be resolved on Windows and MacOS, since neither the `std::numeric_limits< std::chrono::system_clock::duration::rep >` nor the `std::chrono::system_clock::time_point::min( )`/`::max( )` seem to work. At this point my only other ideas are:
- Hard-coding a range that works on all platforms
- Implementing the conversion method to Python datetime objects directly in Python (TBC if this is possible with pybind). This might also have the advantage that a broader range of dates is allowed than when going through a `chrono` intermediate, Python datetime objects allow a range of years from 1-9999 vs. the 1678-2262 range on Linux

If you have any other ideas let me know!

DominicDirkx commented:
Could you re-iterate: where exactly in the code is the overflow-related issue occurring?

larshinueber committed:
additional time_t output

larshinueber committed:
more platform debug output

larshinueber committed:
more debugging output

larshinueber committed:
fix compile error

larshinueber commented:
Contrary to what I was thinking before (that the limit is due to the representation of either the `std::chrono::time_point` or `std::time_t`) I think the issue is platform dependent:
- Linux: The conversion is limited by the representation of time_point (nanoseconds since 1970-01-01, represented as `long long int`), therefore the conversion will overflow outside the 1678-2262 year range
- MacOS & Windows: The conversion fails in the `std::mktime` call in [dateTime.h#L294](https://github.com/larshinueber/tudat/blob/be8e3f85e26f1b1d63610596385ca714a68435ec/include/tudat/astro/basic_astro/dateTime.h#L294), see [Windows log](https://github.com/tudat-team/tudat/actions/runs/15628119960/job/44026158806?pr=332#step:9:5698) and [MacOS log](https://github.com/tudat-team/tudat/actions/runs/15628119960/job/44026158790?pr=332#step:9:7158). It seems that on Windows the `std::mktime` function is only defined in the range of 1970-3000 (see [Microsoft docs](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/mktime-mktime32-mktime64?view=msvc-170#remarks)), while on MacOS `std::mktime` is defined for years >1900 ([SO #1](https://stackoverflow.com/questions/49761269/apple-clang-mktime-issue), [SO#2](https://stackoverflow.com/questions/35300453/am-i-using-tm-mktime-wrong-and-if-not-is-there-a-workaround)).

This is consistent with the tests:
- Linux passes because the `timePoint` method checks for the limits of the `chrono` representation -> year 1621 is detected as being outside the representable range and throws an error, which is the expected behavior in the test
- MacOS passes the tests for years 2023, 2373, 1910 but fails for 1621
- Windows passes the tests for years 2023, 2373 but fails for 1910

If this is the case (again, please let me know if I'm missing something) I would argue for the second implementation option in my previous comment, to (hopefully) circumvent most of these issues and have a consistent behavior across platforms. Happy to discuss of course :)

DominicDirkx commented:
Thanks, that's clear. Based on what you describe, I would say we add the exception handling inside the C++ code, but have it dependent on the system on which the compilation is taking place. For MacOS/Windows, it's then clear to me where to implement this (the std::mktime, as you link in your post). For Linux, where is the check done? I'm guessing this should still be done on Windows/Mac, but on top if it the check should be done on std::mktime (?)

DominicDirkx committed:
Added Windows/Mac specific time test

DominicDirkx commented:
I've pushed an attempt at a system-dependent error check.

larshinueber committed:
move platform dependent code into respective functions

larshinueber committed:
verbose test output, not only on failure

larshinueber committed:
add additional years as corner cases

larshinueber committed:
add verbose flag to ctest

larshinueber committed:
update leap year check

larshinueber committed:
add test case output

larshinueber committed:
exclude bounds when checking for limits

larshinueber committed:
fix test syntax error

larshinueber commented:
@DominicDirkx Thank you! I've incorporated the system-specific code into the check functions and added a few edge test cases.
This should finally work now :)

Before merging I would remove all the logging output and update the docs for the new behavior. Will let you know once this is ready to merge 👍

larshinueber committed:
re-enable all tests, suppress output by default

DominicDirkx commented:
That's super awesome! Merging this now :)
